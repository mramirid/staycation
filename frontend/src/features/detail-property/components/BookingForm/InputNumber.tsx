import { formatCountSuffix, type Suffix } from "@/utils/format";
import { clx } from "@/utils/styling";
import classes from "./input-group.module.scss";

type Props = {
  id: string;
  value: number;
  onChange: (newValue: number) => void;
  min?: number;
  suffix?: Suffix;
};

export default function InputNumber({ min = 1, ...props }: Props) {
  const inputValue = formatCountSuffix(props.value, props.suffix);

  const hasReachedMin = props.value <= min;

  const minus = () => {
    if (hasReachedMin) return;
    props.onChange(props.value - 1);
  };

  const plus = () => {
    props.onChange(props.value + 1);
  };

  return (
    <div className={classes.inputGroup}>
      <button
        className={clx(
          classes.inputGroup__sideIcon,
          "btn btn-error !rounded-l"
        )}
        type="button"
        disabled={hasReachedMin}
        onClick={minus}
        data-testid="btn-minus"
      >
        -
      </button>
      <input
        readOnly
        id={props.id}
        type="text"
        placeholder="Enter a number..."
        className={classes.inputGroup__textInput}
        value={inputValue}
      />
      <button
        className={clx(
          classes.inputGroup__sideIcon,
          "btn btn-success !rounded-r"
        )}
        type="button"
        onClick={plus}
        data-testid="btn-plus"
      >
        +
      </button>
    </div>
  );
}
