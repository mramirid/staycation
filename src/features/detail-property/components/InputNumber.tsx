import { clx } from "@/utils/styling";
import { inRange, isString, isUndefined } from "lodash-es";
import { useState } from "react";

type Props = {
  className?: string;
  name: string;
  value: number;
  onChange: (newValue: number) => void;
  min?: number;
  max?: number;
  suffix?: { value: string; pluralValue?: string };
};

export default function InputNumber({ min = 1, max = 1, ...props }: Props) {
  if (!inRange(props.value, min, max + 1)) {
    throw new Error("The value is out of range");
  }

  const buildSuffix = (currentValue: number) => {
    const { suffix } = props;

    if (isUndefined(suffix)) {
      return "";
    }

    if (isString(suffix.pluralValue) && currentValue > 1) {
      return suffix.pluralValue;
    }
    return suffix.value;
  };

  const [inputValue, setInputValue] = useState(
    [props.value, buildSuffix(props.value)].join(" ")
  );

  const onChange = (newValue: string) => {
    const currentSuffix = buildSuffix(props.value);
    if (isString(currentSuffix)) {
      newValue = newValue.replace(currentSuffix, "");
    }

    const patternNumeric = new RegExp("[0-9]*");
    const isNumeric = patternNumeric.test(newValue);

    if (isNumeric && +newValue <= max && +newValue >= min) {
      props.onChange(+newValue);

      const newSuffix = buildSuffix(+newValue);
      setInputValue([newValue, newSuffix].join(" "));
    }
  };

  const hasReachedMin = props.value === min;
  const hasReachedMax = props.value === max;

  const minus = () => {
    if (hasReachedMin) return;
    onChange((props.value - 1).toString());
  };

  const plus = () => {
    if (hasReachedMax) return;
    onChange((props.value + 1).toString());
  };

  return (
    <div
      className={clx("input-group max-h-fit items-stretch", props.className)}
    >
      <button
        className="input-side-icon btn btn-error !rounded-l font-bold text-2xl h-full"
        type="button"
        disabled={hasReachedMin}
        onClick={minus}
        data-testid="btn-minus"
      >
        -
      </button>
      <input
        type="text"
        min={min}
        max={max}
        name={props.name}
        pattern="[0-9]*"
        className="bg-base-200 text-center focus:outline-none grow min-w-0"
        value={inputValue}
        onChange={(e) => onChange(e.target.value)}
      />
      <button
        className="input-side-icon btn btn-success !rounded-r font-bold text-2xl h-full"
        type="button"
        disabled={hasReachedMax}
        onClick={plus}
        data-testid="btn-plus"
      >
        +
      </button>
    </div>
  );
}
