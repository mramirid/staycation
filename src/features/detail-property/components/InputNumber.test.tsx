import { fireEvent, render, screen } from "@testing-library/react";
import UserEvent from "@testing-library/user-event";
import { useState } from "react";
import InputNumber from "./InputNumber";

function Form() {
  const [value, setValue] = useState<number>(2);

  const handleChange = (newValue: number) => {
    setValue(newValue);
  };

  return (
    <InputNumber
      id="test-input-number"
      value={value}
      min={1}
      suffix={{ value: "day", pluralValue: "days" }}
      onChange={handleChange}
    />
  );
}

describe("<InputNumber />", () => {
  it("Should match with the snapshot", () => {
    const { container } = render(<Form />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Should able to change value", () => {
    render(<Form />);

    const input = screen.getByRole("textbox") as HTMLInputElement;
    fireEvent.change(input, { target: { value: 50 } });

    expect(input.value).toBe("50 days");
  });

  it("Should not be able to change when it has reached the mininmum value", () => {
    render(<Form />);

    const input = screen.getByRole("textbox") as HTMLInputElement;
    fireEvent.change(input, { target: { value: 0 } });

    expect(input.value).toBe("2 days");
  });

  it("Should increment when the plus button is pressed", async () => {
    render(<Form />);

    const plusButton = screen.getByTestId("btn-plus") as HTMLButtonElement;
    await UserEvent.click(plusButton);

    const input = screen.getByRole("textbox") as HTMLInputElement;
    expect(input.value).toBe("3 days");
  });

  it("Should decrement when the minus button is pressed", async () => {
    render(<Form />);

    const plusButton = screen.getByTestId("btn-minus") as HTMLButtonElement;
    await UserEvent.click(plusButton);

    const input = screen.getByRole("textbox") as HTMLInputElement;
    expect(input.value).toBe("1 day");
  });
});
