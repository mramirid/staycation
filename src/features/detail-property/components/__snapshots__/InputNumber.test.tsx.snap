// Vitest Snapshot v1

exports[`<InputNumber /> > Should match with the snapshot 1`] = `
<div
  class="_inputGroup_zgol5_1"
>
  <button
    class="_inputGroup__sideIcon_zgol5_4 btn btn-error !rounded-l"
    data-testid="btn-minus"
    type="button"
  >
    -
  </button>
  <input
    class="_inputGroup__textInput_zgol5_9"
    id="test-input-number"
    placeholder="Enter a number..."
    type="text"
    value="2 days"
  />
  <button
    class="_inputGroup__sideIcon_zgol5_4 btn btn-success !rounded-r"
    data-testid="btn-plus"
    type="button"
  >
    +
  </button>
</div>
`;
